### PR TITLE
Feature 42 comments

### PR DESCRIPTION
**Pull request by kaeline**
+ Users assigned to the tasks get assigned to the projects as well
+ Users mentioned in the comment must be part of the project in order to see the task and the comments
+ Notifcations on a comment left are sent to: admin, task assignees, referenced users. The user leaving the comment is excluded from the notification

https://github.com/Metatavu/lipsanen-project-management-api/issues/42

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+        val task = taskController.find(projectMilestone!!.first, taskId) ?: return@async createNotFound(
+            createNotFoundMessage(TASK, taskId)
+        )
+
+        taskComment.referencedUsers.forEach { referencedUserId ->
+            if (!projectController.hasAccessToProject(task.milestone.project, referencedUserId)) {
+                return@async createBadRequest("Referenced user $referencedUserId is not an assignee of the task")
+            }
+        }
```
Comment (Daniilmass) [inline src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt:84]: So we can not reference users that are not assigned to this task?

File: src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            createNotFoundMessage(TASK_COMMENT, commentId)
+        )
+
+        if (comment.creatorId != userId) return@async createForbidden("Cannot edit not own comment")
+
+        taskComment.referencedUsers.forEach { referencedUserId ->
+            if (!projectController.hasAccessToProject(task.milestone.project, referencedUserId)) {
+                return@async createBadRequest("Referenced user $referencedUserId is not an assignee of the task")
+            }
+        }
```
Comment (Daniilmass) [inline src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt:142]: Same here, I would consider a case when User would want to reference a non-assignee as a possible one. I.e. task assignee (person doing the task) wants to reference their accountant / project manager in a comment. Those people are likely not assigned to this same task

File: src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt (hunk 1 above)
Comment (kaeline) [reply src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt:84]: No i just noticed it's wrong comment. We can't reference user that has no rights for the project

Comment (kaeline) [reply src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt]: Fixed the message

File: src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt (hunk 2 above)
Comment (kaeline) [reply src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/comments/TaskCommentsApiImpl.kt]: Fixed the message

**Review by Daniilmass — APPROVED** [review]

**Review by lauriLukkarinen — APPROVED** [review]

**Pull request merged**